dash/extract-meta.js

```diff
@@ -1,126 +1,737 @@
+    reservedPatterns.forEach(reservedPattern => {
+        if (reservedPattern.test(propName)) {
+            process.stderr.write(
+                `\nERROR: "${propName}" matches reserved word ` +
+                    `pattern: ${reservedPattern.toString()}\n`
+            );
+            failedBuild = true;
+        }
+    });
+    return failedBuild;
+}
+
+function checkDocstring(name, value) {
+    if (
+        !value ||
+        (value.length < 1 && !excludedDocProps.includes(name.split('.').pop()))
+    ) {
+        logError(`\nDescription for ${name} is missing!`);
     }
 }
 
 function docstringWarning(doc) {
-    checkWarn(doc.displayName, doc.description);
+    checkDocstring(doc.displayName, doc.description);
 
-    Object.entries(doc.props).forEach(
-        ([name, p]) => checkWarn(`${doc.displayName}.${name}`, p.description)
+    Object.entries(doc.props).forEach(([name, p]) =>
+        checkDocstring(`${doc.displayName}.${name}`, p.description)
     );
 }
 
-function propError(doc) {
-    for(const propName in doc.props) {
-        reservedPatterns.forEach(reservedPattern => {
-            if (reservedPattern.test(propName)) {
-                process.stderr.write(
-                    `\nERROR: "${propName}" matches reserved word ` +
-                    `pattern: ${reservedPattern.toString()}\n`
-                );
-                failed = true;
-            }
-        });
+function zipArrays(...arrays) {
+    const arr = [];
+    for (let i = 0; i <= arrays[0].length - 1; i++) {
+        arr.push(arrays.map(a => a[i]));
     }
+    return arr;
 }
 
+function cleanPath(filepath) {
+    return filepath.split(path.sep).join('/');
+}
 
-function parseFile(filepath) {
-    const urlpath = filepath.split(path.sep).join('/');
-    let src;
-
-    if (!['.jsx', '.js'].includes(path.extname(filepath))) {
-        return;
-    }
-
+function parseJSX(filepath) {
     try {
-        src = fs.readFileSync(filepath);
-        const doc = metadata[urlpath] = reactDocs.parse(src);
+        const src = fs.readFileSync(filepath);
+        const doc = reactDocs.parse(src);
+        Object.keys(doc.props).forEach(propName =>
+            isReservedPropName(propName)
+        );
         docstringWarning(doc);
-        propError(doc);
+        return doc;
     } catch (error) {
-        writeError(error, filepath);
+        logError(error);
     }
 }
 
-function collectMetadataRecursively(componentPath) {
-    if (ignorePattern.test(componentPath)) {
-        return;
+function gatherComponents(sources, components = {}) {
+    const names = [];
+    const filepaths = [];
+
+    const gather = filepath => {
+        if (ignorePattern && ignorePattern.test(filepath)) {
+            return;
+        }
+        const extension = path.extname(filepath);
+        if (['.jsx', '.js'].includes(extension)) {
+            components[cleanPath(filepath)] = parseJSX(filepath);
+        } else if (filepath.endsWith('.tsx')) {
```

Comment:
Is this 100% consistent? Components are always defined in `.tsx` files, never `.ts` or something like `.react.ts`?

Reply:
Yes, in typescript you can only add jsx elements in tsx files.